sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala

```diff
@@ -858,6 +897,74 @@ trait OrcCommonFunctions extends OrcCodecWritingHelper { self: FilePartitionRead
         throw new IOException(s"Error when processing file splits [${splits.mkString("; ")}]", e)
     }
   }
+
+  protected final def isNeedToSplitDataBlock(
+      curMeta: OrcBlockMetaForSplitCheck,
+      nextMeta: OrcBlockMetaForSplitCheck): Boolean = {
+    if (!nextMeta.typeDescription.equals(curMeta.typeDescription)) {
+      logInfo(s"ORC schema for the next file ${nextMeta.filePath}" +
+        s" schema ${nextMeta.typeDescription} doesn't match current ${curMeta.filePath}" +
+        s" schema ${curMeta.typeDescription}, splitting it into another batch!")
+      return true
+    }
+
+    if (nextMeta.compressionKind != curMeta.compressionKind) {
+      logInfo(s"ORC File compression for the next file ${nextMeta.filePath}" +
+        s" doesn't match current ${curMeta.filePath}, splitting it into another batch!")
+      return true
+    }
+
+    val ret = (nextMeta.requestedMapping, curMeta.requestedMapping) match {
+      case (None, None) => true
+      case (Some(cols1), Some(cols2)) =>
+        if (cols1.sameElements(cols2)) true else false
+      case (_, _) => {
+        false
+      }
+    }
+
+    if (!ret) {
+      logInfo(s"ORC requested column ids for the next file ${nextMeta.filePath}" +
+        s" doesn't match current ${curMeta.filePath}, splitting it into another batch!")
+      return true
+    }
+
+    false
+  }
+
+  protected implicit def toStripe(block: DataBlockBase): OrcStripeWithMeta =
+    block.asInstanceOf[OrcDataStripe].stripeMeta
+
+  protected implicit def toDataStripes(stripes: Seq[DataBlockBase]): Seq[OrcStripeWithMeta] =
+    stripes.map(_.asInstanceOf[OrcDataStripe].stripeMeta)
+
+  protected final def estimateOutputSizeFromBlocks(blocks: Seq[OrcStripeWithMeta]): Long = {
+    // Start with header magic
+    var size: Long = OrcFile.MAGIC.length
+    blocks.foreach { block =>
+      // Account for the size of every stripe
+      size += block.stripeLength
+
+      // Add StripeInformation size in advance which should be calculated in Footer
+      size += OrcTools.sizeOfStripeInformation
+    }
+
+    if (blocks.nonEmpty) {
+      // Add the first orc file's footer length to cover ORC schema and other info.
+      // Here uses the file footer size from the OrcPartitionReaderContext of the first
+      // stripe as the worst-case.
+      size += blocks.head.ctx.fileTail.getPostscript.getFooterLength
+    }
+
+    // Per ORC v1 spec, the size of Postscript must be less than 256 bytes.
+    size += 256
+    // And the single-byte postscript length at the end of the file
+    size += 1
+    // Add in a bit of fudging in case the whole file is being consumed and
+    // our codec version isn't as efficient as the original writer's codec.
+    size + OrcTools.INEFFICIENT_CODEC_BUF_SIZE
```

Comment:
so this is sufficient size, you didn't run into footer sizes growing after combining? I ran into that in parquet so want to make sure.  What is the max combined you have tested?  does orc basically have fixed sizes for everything?

Reply:
No, I tried at most 128 MB combining size and 3TB input data during my NDS runs on spark2a with data `orc_sf3k_decimal_zstd`.

I believe this is OK because it is not a new algorithm, even not a new function, but is from the coalescing code. I just moved original one to this parement trait, then it can be called by both the multithreaded and coalescing readers.

Reply:
coalescing and multithreaded combining are not the same thing.  I had to add extra space in the parquet side due to footer sizes changing